### PR TITLE
Managed node groups userdata fix and kms support

### LINKED ISSUE #428: [Bug]: Launch template error causing EC2 nodes to fail joining the node group properly
### Welcome to Amazon EKS Blueprints!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/aws-ia/terraform-aws-eks-blueprints/issues) and didn't find any.

### Amazon EKS Blueprints Release version

34921232356fb69b5a41094246d488c37c596d97

### What is your environment, configuration and the example used?

- Terraform 1.1.7

```hcl
module "eks_blueprints" {
  source  = "github.com/aws-ia/terraform-aws-eks-blueprints?ref=34921232356fb69b5a41094246d488c37c596d97"

  org               = var.org
  tenant            = var.tenant
  environment       = var.environment
  zone              = var.zone
  tags              = var.tags

  // NOTE: Add-ons are not updated automatically when the cluster is updated.
  // You will need to do this at the appropriate time for each addon, if it is required.
  // See: https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/main/docs/add-ons/index.md
  terraform_version = var.terraform_version

  # KMS
  cluster_kms_key_additional_admin_arns = var.cluster_kms_key_additional_admin_arns

  # Auth
  map_roles         = var.map_roles
  application_teams = var.application_teams
  platform_teams    = var.platform_teams
  
  # EKS Cluster VPC and Subnets
  vpc_id             = local.vpc_id
  private_subnet_ids = local.private_subnet_ids
  //public_subnet_ids  = local.public_subnet_ids

  # EKS CONTROL PLANE VARIABLES
  cluster_version                  = var.cluster_version
  cluster_endpoint_private_access  = false
  cluster_endpoint_public_access   = true

  # EKS MANAGED NODE GROUPS
  managed_node_groups = {
  ondemand_v1 = {
    node_group_name = "ondemand_v1"

    create_launch_template  = true
    launch_template_os      = "amazonlinux2eks"

    instance_types = ["t3a.large"]
    subnet_type    = "private"
    ami_type       = "AL2_x86_64"
    capacity_type  = "ON_DEMAND"
    disk_size      = 50

    desired_size    = 1
    min_size        = 1
    max_size        = 4
    max_unavailable = 1
  }
}
}
```

### What did you do and What did you see instead?

This will produce 2 node groups, instead of 1.

CleanShot 2022-04-19 at 14 32 52

The node group created by the AWS blueprints module [aws-eks-managed-node-groups](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/modules/aws-eks-managed-node-groups) appears to be correct (the user data looks good), but it also appears to be unused.

Instead, it appears that the underlying https://github.com/terraform-aws-modules/terraform-aws-eks is ALSO creating a launch template and that template is being used by the instance, but is broken.

This template https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/templates/linux_user_data.tpl is being used and is creating multiple bootstrap lines, which is causing the script to fail and therefore causing the node group to fail.

This is what the incorrect script looks like:

CleanShot 2022-04-19 at 14 40 14

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by vara-bonthu**
### What does this PR do?

- Fix #428 #373
- Managed Node groups supports multiple disk device configuration with Custom KMS key id 
- Examples updated to add additional sec group rules. These are added to ensure the Add-ons working correctly on required ports. Additional comments added to restrict that to specific port based on your needs
- Managed Node groups example updated to show how to use custom ami and ami type scenarios. 
- Service ipv4 and ipv6 cidr variables introduced

### More

- [x] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [ ] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [x] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators
- [x] E2E Test successfully complete before merge?

### CODE REVIEW

File: locals.tf [hunk 1]
```diff
@@ -52,6 +54,10 @@ locals {
     iam_role_path                 = var.iam_role_path
     iam_role_permissions_boundary = var.iam_role_permissions_boundary
 
+    # Service IPv4/IPv6 CIDR range
+    service_ipv6_cidr = local.service_ipv6_cidr
```
Comment (bryantbiggs) [inline locals.tf:58]: why add variables just to immediately convert them to locals? can't we just use them as variables as is?

File: modules/aws-eks-managed-node-groups/locals.tf [hunk 2]
```diff
@@ -9,14 +9,13 @@ locals {
...
+    force_update_version     = null
 
     desired_size    = "3"
     max_size        = "3"
     min_size        = "1"
     max_unavailable = "1"
-
-    disk_size = 50
-    disk_type = "gp3"
+    disk_size       = 50
```
Comment (bryantbiggs) [inline modules/aws-eks-managed-node-groups/locals.tf:18]: there should be `disk_size` or `block_device_mappings`, not both

File: locals.tf (hunk 1 above)
Comment (vara-bonthu) [reply locals.tf:58]: Good point. Will update that.

File: modules/aws-eks-managed-node-groups/locals.tf (hunk 2 above)
Comment (vara-bonthu) [reply modules/aws-eks-managed-node-groups/locals.tf:18]: `disk_size` variable is used for Managed Node groups without Launch Templates.

Comment (bryantbiggs) [reply modules/aws-eks-managed-node-groups/locals.tf:18]: oh I see - this is like a set of "defaults". I would reconsider this approach - it expects users to know where to dive into the source code to find out where the default value is coming from

**Review by kcoleman731 — COMMENTED** [review]
Looking good.

File: docs/index.md [hunk 3]
```diff
@@ -4,9 +4,9 @@
 
 Welcome to Amazon EKS Blueprints for Terraform!
 
-This repository contains a collection of Terraform modules that aim to make it easier and faster for customers to adopt [Amazon EKS](https://aws.amazon.com/eks/). 
+This repository contains a collection of Terraform modules that aim to make it easier and faster for customers to adopt [Amazon EKS](https://aws.amazon.com/eks/).
 
-## WHat is EKS Blueprints 
+## WHat is EKS Blueprints
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply docs/index.md:9]: will update that 👍🏼

File: modules/launch-templates/locals.tf [hunk 4]
```diff
@@ -22,7 +22,7 @@ locals {
     block_device_mappings = {
       device_name           = "/dev/xvda"
       volume_type           = "gp3" # The volume type. Can be standard, gp2, gp3, io1, io2, sc1 or st1 (Default: gp3).
-      volume_size           = "200"
+      volume_size           = "100"
```
Comment (bryantbiggs) [inline modules/launch-templates/locals.tf:25]: this could potentially have a significant impact to users - should we leave this as is for now?

Comment (vara-bonthu) [reply modules/launch-templates/locals.tf:25]: Did you mean to change this back to 200 so that existing customers wont be impacted?

Comment (vara-bonthu) [reply modules/launch-templates/locals.tf]: Updated

**Review by bryantbiggs — APPROVED** [review]

**Review by askulkarni2 — APPROVED** [review]
LGTM!

**Comment by alicancakil** [comment]
thank you so much for fixing this. 

This saved me: `launch_template_os     = "bottlerocket"`

**Pull request merged**